cli/src/main/java/org/jboss/as/cli/operation/impl/DefaultOperationCandidatesProvider.java

```diff
@@ -210,6 +212,193 @@ public CommandLineCompleter getCompleter() {
+        public PropertyVisibility(boolean all, List<Property> propList,
+                Set<String> presentProperties, String radical) {
+            this.all = all;
+            this.propList = propList;
+            this.presentProperties = Collections.unmodifiableSet(presentProperties);
+            this.radical = radical;
+
+            // Create some maps to speedup lookup done further down the road.
+            Map<String, Set<String>> requiresMap = new HashMap<>();
+            Map<String, Set<String>> alternativesMap = new HashMap<>();
+            for (Property prop : propList) {
+                // Present property, hides alternatives.
+                if (presentProperties.contains(prop.getName())) {
+                    if (prop.getValue().hasDefined(Util.ALTERNATIVES)) {
+                        List<ModelNode> alternatives = prop.getValue().get(Util.ALTERNATIVES).asList();
+                        for (ModelNode alt : alternatives) {
+                            invalidProperties.add(alt.asString());
+                        }
+                    }
+                }
+                if (prop.getValue().hasDefined(Util.ALTERNATIVES)) {
+                    Set<String> set = new HashSet<>();
+                    alternativesMap.put(prop.getName(), set);
+                    List<ModelNode> alternatives = prop.getValue().get(Util.ALTERNATIVES).asList();
+                    for (ModelNode alt : alternatives) {
+                        set.add(alt.asString());
+                    }
+                }
+                if (prop.getValue().hasDefined(Util.REQUIRES)) {
+                    Set<String> set = new HashSet<>();
+                    requiresMap.put(prop.getName(), set);
+                    List<ModelNode> requires = prop.getValue().get(Util.REQUIRES).asList();
+                    for (ModelNode req : requires) {
+                        set.add(req.asString());
+                    }
+                }
+            }
+            // At this point all alternatives have been hidden.
+            // All present non hidden properties must have their requires computed.
+            for (Property prop : propList) {
+                if (presentProperties.contains(prop.getName())) {
+                    if (!invalidProperties.contains(prop.getName())) {
+                        Set<String> requires = requiresMap.get(prop.getName());
+                        if (requires != null) {
+                            for (String req : requires) {
+                                // A require could be hidden because it is in alternatives with a present property.
+                                if (!invalidProperties.contains(req)) {
+                                    Set<String> requiring = requiredBy.get(req);
+                                    if (requiring == null) {
+                                        requiring = new HashSet<>();
+                                        requiredBy.put(req, requiring);
```

Comment:
Please ignore this if I'm wrong; i.e. it's a lot of code and there's no need to waste your time explaining how I'm wrong. ;) But I don't see any reads of the Set<String> values stored in requiredBy. I just see at L355 "return requiredBy.get(prop.getName()) != null;". So it seems that requiredBy could just be a Set<String> and not a map.

Reply:
Yes, no need for a map. I think that I used it to help debug. Replaced.